oracle and randomness overhaul (#551)
loomnetwork/cryptozombie-lessons: en/20/01.md

---
title: Chainlink Data Feeds Introduction
actions: ['checkAnswer', 'hints']
requireLogin: true
material:
  editor:
    language: sol
    startingCode:
      "PriceConsumerV3.sol": |
        // Start here
    answer: |
      pragma solidity ^0.6.7;

      contract PriceConsumerV3 {

      }
---

Now, let's suppose you're building a DeFi dapp, and want to give your users the ability to withdraw ETH worth a certain amount of USD. To fulfill this request, your smart contract (for simplicity's sake we'll call it the "caller contract" from here onwards) must know how much one Ether is worth.

And here's the thing: a JavaScript application can easily fetch this kind of information, making requests to the Binance public API (or any other service that publicly provides a price feed). But, a smart contract can't directly access data from the outside world. 

> Now we could build a javascript application ourselves, but then we are introducing a centralized point of failure! We also couldn't just pull from the Binance API, because that would be a centralized point of failure!

Reviewer: ```suggestion
Now we could build a JavaScript application ourselves, but then we are introducing a centralized point of failure! We also couldn't just pull from the Binance API, because that would be a centralized point of failure!
```

So what we want to do, is get our data from both a **decentralized oracle network (DON)** and **decentralized data sources**. 

Chainlink is a framework for decentralized oracle networks (DONs), and is a way to get data in from multiple sources across multiple oracles. This DON aggregates data in a decentralized manner and places it on the blockchain in a smart contract (often referred to as a "price reference feed" or "data feed") for us to read from. So all we have to do, is read from a contract that the Chainlink network is constantly updating for us!

> Using Chainlink Data Feeds is a way to cheaply, more accurately, and with more security gather data from the real world in this decentralized context. Since the data is coming from multiple sources, multiple people can partake in the ecosystem and it becomes **even cheaper than running even a centralized oracle.** The Chainlink network uses a system called [Off-Chain Reporting](https://docs.chain.link/docs/off-chain-reporting/) to reach a consensus on data off-chain, and report the data in a cryptographically proven single transaction back on chain for users to digest. 

Reviewer: ```suggestion
Using Chainlink Data Feeds is a way to cheaply, more accurately, and with more security gather data from the real world in this decentralized context. Since the data is coming from multiple sources, multiple people can partake in the ecosystem and it becomes **even cheaper than running even a centralized oracle.** The Chainlink network uses a system called <a href="https://docs.chain.link/docs/off-chain-reporting/" target=_new>Off-Chain Reporting</a> to reach a consensus on data off-chain, and report the data in a cryptographically proven single transaction back on-chain for users to digest. 
```

> You can then make protocols like [Synthetix](https://synthetix.io/), [Aave](https://aave.com/), and [Compound](https://compound.finance/) with this!

Reviewer: ```suggestion
You can then make protocols like <a href="https://synthetix.io/" target=_new>Synthetix</a>, <a href="https://aave.com/)" target=_new>Aave</a>, and <a href="https://compound.finance/)" target=_new>Compound</a> with this!
```

> <img src="./img/price-feed-diagram.png" alt="Chainlink Decentralized Oracle Network" width="469">

Reviewer: ```suggestion
<img src="/course/static/image/lesson-20/price-feed-diagram.png" alt="Chainlink Decentralized Oracle Network" width="469">
```

> You can see visualizations of [some of these decentralized oracle networks (DONs) here.](https://data.chain.link/)

Reviewer: ```suggestion
<a href="(https://data.chain.link/" target=_new>some of these DONs here</a>
You can see visualizations of <a href="(https://data.chain.link/" target=_new>some of these DONs here</a>.
```

> We'll go into exactly how these networks tick in later sessions. 

Reviewer: ```suggestion
We'll go into exactly how these networks tick in later lessons. 
```

So, let's learn how to read from one of these data feeds!

The first thing we want to do, is start our contract and import the Chainlink code. 

## Putting it to the Test

Some of this should be pretty familiar to you, the only real different part is the implications!

1. Declare our `pragma` version at the top of the file to `^0.6.7` (check previous lessons' files for the syntax).

> 2. Declare a new contract, `PriceConsumerV3`. Leave the body of the contract empty for now.

Reviewer: ```suggestion
2. Now, you would want to declare a new contract named `PriceConsumer`. The current version of the Chainlink aggregator interface is v3. However, as Chainlink adds more features to the interface, this version change over time. When you name your contract, let's append this version to the name itself for later reference. Declare a new contract named `PriceConsumerV3`, and leave the body of the contract empty for now.
```
